docs/pallas/design.md

```diff
@@ -0,0 +1,317 @@
+
+Note that these APIs are still experimental and subject to change.
+
+### Reference types
+
+Let’s look at an example Pallas program for adding two vectors:
+
+```python
+import jax
+import jax.numpy as jnp
+from jax.experimental import pallas as pl
+
+def add_kernel(x_ref, y_ref, o_ref):
+  # In this code, `x_ref`, `y_ref` and `o_ref` are (8,)-shaped `Ref`s
+  x = x_ref[:]
+  y = y_ref[:]
+  o_ref[:] = x + y
+x, y = jnp.arange(8), jnp.arange(8, 16)
+add = pl.pallas_call(add_kernel, out_shape=jax.ShapeDtypeStruct((8,), jnp.int32))
+add(x, y)
+```
+
+Unlike a regular JAX program, `add_kernel` does not receive immutable array arguments. Instead, it’s provided with references that can be read from and updated in-place using NumPy-like syntax. `Ref`s are not a Pallas-specific concept – they were introduced to JAX to represent stateful computations. However, we can leverage them when writing kernels that operate on mutable memory too.
+
+Pallas kernels not only receive `Ref`s corresponding to the inputs to the kernel, but also receive `Ref`s for the outputs as well (specified in `pallas_call` via `out_shape`). `Ref`s are special types that cannot be passed into the usual set of JAX primitives without being read from first. When you read from a `Ref` you get a JAX `Array` type out, and you must write an `Array` into a `Ref`.
+
+#### Reading from/writing into Refs
+
+Reading from a `Ref` corresponds to loading an array into the lowest level of the memory hierarchy (L1-cache on GPU and vector registers on TPU). Writing into a `Ref` is analogous.
+
+```python
+def f(x_ref, o_ref):
+  # Using vanilla Python indexing
+  x = x_ref[0, 2:5, :]
+  # Or via Numpy advanced int indexing
+  o_ref[jnp.arange(3), :] = x
+
+Note that in order to use NumPy advanced int indexing, you need to broadcast the indices against each other into the desired multidimensional shape:
+def f(x_ref):
+  # Assume x_ref is (8, 4) and we want to read out a (2, 3) slice
+  x = x_ref[jnp.arange(2)[..., None], jnp.arange(3)[None, ...]]
+```
+
+Writing to `Ref`s can be done via analogous `__setitem__` style indexing.
+
+Other forms of indexing (for example, dynamic slicing) can be done via `pallas.load` and `pallas.store`, new JAX primitives designed to make loading from/storing into memory easier. We’ll discuss these new primitives later.
+
+### Extending JAX with new Pallas primitives
+
+Because JAX was designed with HLO in mind, the set of JAX primitives closely mirrors the set of HLO operations. Targeting a new compiler (e.g. Triton or Mosaic) means we might need to supplement JAX’s primitives with new ones specific to the new compiler. At the same time, we may not be able to lower all JAX primitives, so we need to restrict it to a subset.
+
+Because Pallas was initially designed with Triton in mind, we offer a set of new primitives with the Triton programming model in mind. As we’ll show later, we can lower these primitives to Mosaic as well.
```

Comment:
word repetition:
```suggestion
Because Pallas was initially designed with Triton in mind, we offer a set of new primitives targeting the Triton programming model. As we’ll show later, we can lower these primitives to Mosaic as well.
```